bindings.c

```diff
@@ -4092,6 +4292,363 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
 	return rv;
 }
 
+/*
+ * Find the process pid from cgroup path.
+ * eg:from /sys/fs/cgroup/cpu/docker/containerid/cgroup.procs to find the process pid.
+ * @pid_buf : put pid to pid_buf.
+ * @dpath : the path of cgroup. eg: /docker/containerid or /docker/containerid/child-cgroup ...
+ * @depth : the depth of cgroup in container.
+ * @sum : return the number of pid.
+ * @cfd : the file descriptor of the mounted cgroup. eg: /sys/fs/cgroup/cpu
+ */
+static int calc_pid(char ***pid_buf, char *dpath, int depth, int sum, int cfd)
+{
+	DIR *dir;
+	int fd;
+	struct dirent *file;
+	FILE *f = NULL;
+	size_t linelen = 0;
+	char *line = NULL;
+	int pd;
+	char *path_dir, *path;
+	char **pid;
+
+	/* path = dpath + "/cgroup.procs" + /0 */
+	path = (char *)malloc(strlen(dpath) + 20);
+	if (!path) {
+		lxcfs_error("Failed to malloc %s in calc_pid!\n", dpath);
+		return sum;
+	}
+
+	strcpy(path, dpath);
+	fd = openat(cfd, path, O_RDONLY);
+	if (fd < 0)
+		goto out;
+
+	dir = fdopendir(fd);
+	if (dir == NULL)
+		goto out;
```

Comment:
If `fdopendir()` fails you still need to cleanup the file descriptor `fd`. Currently, you're leaking it below.

Reply:
Yes, you're right.
I will rewrite them.

Reply:
Make sure that you only close the `fd` via `close(fd)` when `fdopendir()` fails. After `fdopendir()` has returned successfully you can only close `dir` via `closedir(dir)`.

Reply:
👍 
Got it！